lib/Dialects/LinalgTransform/Transforms/PDL.cpp

```diff
@@ -59,6 +63,240 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
+                                          PatternRewriter &rewriter) {
+  if (succeeded(haveIdenticalBodiesImpl(linalgOp, genericLinalgModelOp)))
+    return success();
+  // TODO: haveEquivalentBodiesImpl, see e.g.
+  // https://gist.github.com/nicolasvasilache/39e89e18c46e02335c16db6ec20a07e3
+  return failure();
+}
+
+/// Succeed when `linalgOp` and `linalgModelOp` are deemed equivalent.
+static LogicalResult isEquivalentToOpImpl(LinalgOp linalgOp,
+                                          LinalgOp linalgModelOp,
+                                          PatternRewriter &rewriter) {
+  // If basic properties do not match, return failure.
+  if (linalgOp.inputs() != linalgModelOp.inputs() ||
+      linalgOp.outputs() != linalgModelOp.outputs() ||
+      linalgOp.indexing_maps() != linalgModelOp.indexing_maps() ||
+      linalgOp.iterator_types() != linalgModelOp.iterator_types())
+    return failure();
+
+  // Build the block and go perform a body comparison.
+  {
+    // createBlock moves the insertion point, scope it in an RAII block.
+    OpBuilder::InsertionGuard guard(rewriter);
+    Region &r = linalgModelOp->getRegion(0);
+    Block *bodyBlock = rewriter.createBlock(
+        &r, r.end(), linalgOp.getBlock()->getArgumentTypes(),
+        llvm::to_vector<4>(
+            llvm::map_range(linalgOp.getBlock()->getArguments(),
+                            [](Value v) { return v.getLoc(); })));
+    ImplicitLocOpBuilder b(linalgModelOp.getLoc(), rewriter);
+    auto regionBuilder = linalgModelOp.getRegionBuilder();
+    regionBuilder(b, *bodyBlock);
+  }
+
+  return haveEquivalentBodies(linalgOp, linalgModelOp, rewriter);
+}
+
+/// Check whether the unique Operation* `op` stored in `value` (assumed) is
+/// equivalent to the unique StringRefAttr operation name passed in
+/// `constantParams`.
+/// Equivalence is achieved when either:
+///   1. `op` has the name stored in `constantParams`.
+///   2. `op` and `constantParams` are both linalg ops and their structured
+///      interfaces as well as their bodies are equivalent.
+///      Structured interfaces equivalence is a simple attribute level check.
+///      Body equivalence is more involved and currently limited:
+///        a. the current impl constructs an instance of the op whose name is
+///           specified in `constantParams` and checks for exact body equality.
+///        b. a more advanced version would "subtract" the bodies and fold, cse
+///           and canonicalize to fixed point. If the result is "all zeros",
+///           then the bodies would be equivalent (really isomorphic).
+///   3. other cases TBD (e.g. vector.generic when available).
+static LogicalResult isEquivalentToOp(PDLValue value, ArrayAttr constantParams,
+                                      PatternRewriter &rewriter) {
+  auto maybeOp = value.dyn_cast<Operation *>();
+  if (!maybeOp)
+    return failure(); // TODO: notifyMatchFailure needs an Operation* handle.
+  Operation *op = *maybeOp;
```

Comment:
I am trying this right now :) This is supposed to be `Operation *op = maybeOp;`, right?

Reply:
Fantastic, the more the merrier.
Next, for my part, I'll connect this to the tuner.